### PR TITLE
allow evaluation with backbone different from Resnet

### PR DESCRIPTION
**Pull request by Ori226**
Support evaluation of the other available back-ends. Please note that the back-ends are identify by the file name.

### CODE REVIEW

File: keras_retinanet/bin/evaluate.py [hunk 1]
```diff
@@ -114,8 +112,22 @@ def main(args=None):
...
+    elif 'densenet' in os.path.basename(args.model):
+        from keras_retinanet.models.densenet import custom_objects
+    elif 'vgg' in os.path.basename(args.model):
+        from keras_retinanet.models.vgg import custom_objects
+    else:
+        raise NotImplementedError('Un-identified backbone. Not implemented.')
+
     # load the model
     print('Loading model, this may take a second...')
+
```
Comment (de-vri-es) [inline keras_retinanet/bin/evaluate.py]: Could you remove this blank line?

Comment (Ori226) [reply keras_retinanet/bin/evaluate.py]: yes

**Comment by hgaiser** [comment]
Hmm the model name might not necessarily include the backbone. I'd be more in favour of adding an argument to define what backbone to use. Even better would be if we can find a way to encode in the model what backbone it contains, but I'm not sure how to do this.

**Comment by rodrigo2019** [comment]
I'm working in a similar thing, here is a adapted example extracted from my code
I'm creating a callback that saves information like the losses value on a saved model 

```python
#write information
import h5py
f = h5py.File('../snapshots/tiny_darknet_pascal_104.h5','a')
f['Informations/backbone'] = 'tiny_darknet'
f.close()

#test model
import sys
sys.path.append("..")
from keras.models import load_model
from keras_retinanet.models.darknet import custom_objects

model = load_model('../snapshots/tiny_darknet_pascal_01.h5', custom_objects=custom_objects)

#print backbone information    
f = h5py.File('../snapshots/tiny_darknet_pascal_104.h5', 'r')
print(f['Informations/backbone'].value)
f.close()
```

**Comment by de-vri-es** [comment]
One problem with that is that we'd need to know the path of the saved snapshots. While they are predictable, duplicating the logic to produce the file name is not very nice. It also requires re-opening the file to append data. I'm not sure how expensive that is with h5py.

Ideally we would tell keras to add some extra metadata, but save/load_model() doesn't currently seem to support that. Plus, load_model already needs to have the custom objects dictionary, so it can't use any metadata from the file to form the dictionary. That is solvable, but for now I think a command line flag is the best solution.

File: keras_retinanet/bin/evaluate.py [hunk 2]
```diff
@@ -113,6 +113,7 @@ def main(args=None):
 
     # create the generator
     generator = create_generator(args)
+    custom_objects = custom_objects_importer(args.backbone)
```
Comment (hgaiser) [inline keras_retinanet/bin/evaluate.py]: Why as `custom_objects_importer`? It's more simple to import it as `custom_objects` and do : 
```python
model = keras.models.load_model(args.model, custom_objects=custom_objects(args.backbone))
```

File: keras_retinanet/bin/evaluate.py [hunk 3]
```diff
@@ -89,6 +89,7 @@ def parse_args(args):
     parser.add_argument('--iou-threshold',   help='IoU Threshold to count for a positive detection (defaults to 0.5).', default=0.5, type=float)
     parser.add_argument('--max-detections',  help='Max Detections per image (defaults to 100).', default=100, type=int)
     parser.add_argument('--save-path',       help='Path for saving images with detections.')
+    parser.add_argument('--backbone', help='Backbone model used by retinanet.', default='resnet50', type=str)
```
Comment (hgaiser) [inline keras_retinanet/bin/evaluate.py]: `type=str` is the default, so can be omitted.

Also, please align on the `help` arg :)

File: keras_retinanet/bin/evaluate.py [hunk 4]
```diff
@@ -116,7 +117,7 @@ def main(args=None):
 
     # load the model
     print('Loading model, this may take a second...')
-    model = keras.models.load_model(args.model, custom_objects=custom_objects)
+    model = keras.models.load_model(args.model, custom_objects=custom_objects(args.backbone))
```
Comment (hgaiser) [inline keras_retinanet/bin/evaluate.py]: I added a utility function for this: https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/models/__init__.py#L5

Comment (Ori226) [reply keras_retinanet/bin/evaluate.py]: @hgaiser  correct me if I'm wrong, but after the update you made, my PR is no longer needed. right?

Comment (de-vri-es) [reply keras_retinanet/bin/evaluate.py]: Err, it seems they did.

**Pull request closed without merging**